Optimize ArcContainer.getContexts() used during client proxy invocation

mkouba opened the pull request:
This pull request supersedes https://github.com/quarkusio/quarkus/pull/27804.

mkouba committed:
Optimize ArcContainer.getContexts() used during client proxy invocation

Sanne commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java:
```diff
@@ -149,6 +153,10 @@ public ArcContainerImpl(CurrentContextFactory currentContextFactory) {
         }
 
+        List<InjectableContext> requestContexts = contexts.get(RequestScoped.class);
+        this.requestContextSingleton = requestContexts.size() == 1 ? Collections.singletonList(requestContexts.get(0)) : null;
```
Why would you store `null` when you could store the list?

mkouba replied:
Sorry, that was an old version...

mkouba committed:
Introduce the idea of Contexts abstraction

- inspired by the pull request from @Sanne:
https://github.com/quarkusio/quarkus/pull/27804

mkouba commented:
@Sanne I've added the `Contexts` abstraction which is built on top of your idea of `ContextsMap`.

mkouba commented:
I'll run the benchmarks tomorrow...

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:
```diff
@@ -0,0 +1,158 @@
+                }
+            };
+            this.size = 3 + mapSize;
+            Set<Class<? extends Annotation>> all = new HashSet<>(contexts.keySet());
```
If `contexts` are not empty, but do _not_ contain a custom implementation of the request context, then `all` won't include `RequestScoped.class` I think?

mkouba replied:
You're right, that's an oversight... it was there a few iterations earlier.

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:
```diff
@@ -0,0 +1,158 @@
+                    return Collections.emptyList();
+                }
+            };
+            this.size = 3 + mapSize;
```
This looks to me like `mapSize` is always 1 more than necessary (and the last element of `keys` and `values` is always `null`), and `size` is only accurate if there's no custom implementation of the request context.

IMHO, `mapSize` should be just `contexts.size()`, and `size` should be `2 + (contexts.containsKey(RequestScoped.class) ? 0 : 1) + mapSize`.

mkouba replied:
Yep, an oversight again..

mkouba replied:
Thanks for noticing! I will add some tests...

Sanne commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:
```diff
@@ -0,0 +1,158 @@
+                            return values[i];
+                        }
+                    }
+                    return Collections.emptyList();
```
are you sure you want to associate an emptyList to any type which is not in the map?
In my implementation I made sure to differentiate, so I could return `null` - but perhaps I was going too far, as I just wanted to emulate exactly the same semantics as you had before.

So perhaps it's fine and perhaps it doesn't matter, but wanted to call out the semantic difference.

mkouba replied:
`ArcContainer.getContexts(Class<? extends Annotation>)` never returns `null` so I think that an empty list is ok.

Ladicek commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Contexts.java:
```diff
@@ -0,0 +1,158 @@
+
+        this.applicationContextSingleton = Collections.singletonList(applicationContext);
+        this.singletonContextSingleton = Collections.singletonList(singletonContext);
+        // request singleton is only set if there is only one context registered for @RequestScoped (99% of cases)
```
This comment is not true? It seems to me that `requestContextSingleton` is always set to a non-`null` value. (And that seems correct, from what I can tell. The comment is probably just stale?)

mkouba replied:
Yes, it is :facepalm:

Ladicek reviewed (approved):
LGTM, modulo one stale comment :-)

mkouba commented:
FTR the [`RequestScopedProxyInvocationBenchmark`](https://github.com/mkouba/arc-benchmarks/blob/master/src/main/java/io/quarkus/arc/benchmarks/RequestScopedProxyInvocationBenchmark.java) shows ~32% improvement.
